src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs

```diff
@@ -0,0 +1,126 @@
+        {
+            ValidateContent(content);
+            Debug.Assert(content.Headers.ContentType != null);
+            Encoding? sourceEncoding = JsonContent.GetEncoding(content.Headers.ContentType.CharSet);
+
+            return ReadFromJsonAsyncCore<T>(content, sourceEncoding, options ?? JsonContent.s_defaultSerializerOptions, cancellationToken);
+        }
+
+        private static async Task<object?> ReadFromJsonAsyncCore(HttpContent content, Type type, Encoding? sourceEncoding, JsonSerializerOptions? options, CancellationToken cancellationToken)
+        {
+            Stream contentStream = await content.ReadAsStreamAsync().ConfigureAwait(false);
+
+            // Wrap content stream into a transcoding stream that buffers the data transcoded from the sourceEncoding to utf-8.
+            if (sourceEncoding != null && sourceEncoding != Encoding.UTF8)
+            {
+                using Stream transcodingStream = new TranscodingReadStream(contentStream, sourceEncoding);
+                return await JsonSerializer.DeserializeAsync(transcodingStream, type, options, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                using (contentStream)
+                {
+                    return await JsonSerializer.DeserializeAsync(contentStream, type, options, cancellationToken).ConfigureAwait(false);
+                }
+            }
+        }
+
+        private static async Task<T> ReadFromJsonAsyncCore<T>(HttpContent content, Encoding? sourceEncoding, JsonSerializerOptions? options, CancellationToken cancellationToken)
+        {
+            Stream contentStream = await content.ReadAsStreamAsync().ConfigureAwait(false);
+
+            // Wrap content stream into a transcoding stream that buffers the data transcoded from the sourceEncoding to utf-8.
+            if (sourceEncoding != null && sourceEncoding != Encoding.UTF8)
+            {
+                using Stream transcodingStream = new TranscodingReadStream(contentStream, sourceEncoding);
+                return await JsonSerializer.DeserializeAsync<T>(transcodingStream, options, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                using (contentStream)
+                {
+                    return await JsonSerializer.DeserializeAsync<T>(contentStream, options, cancellationToken).ConfigureAwait(false);
+                }
+            }
+        }
+
+        private static void ValidateContent(HttpContent content)
+        {
+            if (content == null)
+            {
+                throw new ArgumentNullException(nameof(content));
+            }
+
+            string? mediaType = content.Headers.ContentType?.MediaType;
+
+            if (mediaType == null)
+            {
+                throw new NotSupportedException(SR.ContentTypeNotSupported);
+            }
+
+            mediaType = mediaType.ToLower();
```

Comment:
This `ToLower` overload is culture-aware.  Is that desired?

Also, I assume the vast, vast majority of media types are already going to be lowercased, so we don't need to be concerned about this allocating?

Reply:
I was using `ToLower()` here in order to get case-insensitive comparisons when using  `mediaType.SequenceEquals` on a method where this is passed as a `ROS<char>` https://github.com/dotnet/runtime/pull/33459/files#diff-75e20d2a014e7c134181d017ed9f92d6R96

Maybe this is not the best way to achieve case insensitive?

Reply:
You can also achieve case-insensitivity by doing the various comparisons you're doing as case-insensitive.

My question was more about how the lowercasing was being done, though (same question would apply to how the case-insensitive comparison is done).  This ToLower factors in CultureInfo.CurrentCulture, as culture influences casing.  Other ToLower overloads, like ToLowerInvariant, do not.

@GrabYourPitchforks, what's the right thing to do here?

Reply:
Ok, I removed `ToLower` and now perform case-insensitive comparisons instead. Culture awareness was not desired.

Reply:
To @stephentoub's earlier question: we don't actually have an API that performs case insensitive comparisons the "correct" way per any applicable RFC. That's partly what today's earlier ASCII API review was about. :)

Using `OrdinalIgnoreCase` is the closest thing we have to correct behavior.